src/lib.rs

```diff
@@ -376,6 +551,99 @@ impl Contract {
             .create_community_account(new_community.handle.clone())
     }
 
+    #[payable]
+    pub fn edit_proposal(
+        &mut self,
+        id: ProposalId,
+        body: VersionedProposalBody,
+        labels: HashSet<String>,
+    ) {
+        near_sdk::log!("edit_proposal");
+        require!(
+            self.is_allowed_to_edit_proposal(id, Option::None),
+            "The account is not allowed to edit this proposal"
+        );
+        let editor_id = env::predecessor_account_id();
+        let mut proposal: Proposal =
+            self.proposals.get(id).unwrap_or_else(|| panic!("Proposal id {} not found", id)).into();
+
+        let proposal_body = body.clone().latest_version();
+
+        require!(
+            self.has_moderator(editor_id.clone())
+                || editor_id.clone() == env::current_account_id()
+                || (is_draft(&proposal.snapshot.body.clone().latest_version().timeline)
+                    && (is_empty_review(&proposal_body.timeline)
+                        || is_draft(&proposal_body.timeline))),
```

Comment:
It is worth implementing these functions as methods of Timeline enum instead of free-form functions.